File: sqlite-android-compiler/src/main/java/com/alecstrong/sqlite/android/MarshalSpec.java
package com.alecstrong.sqlite.android;

import com.alecstrong.sqlite.android.model.Column;
import com.alecstrong.sqlite.android.model.Table;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.FieldSpec;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeSpec;
import com.squareup.javapoet.TypeVariableName;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.Map;
import javax.lang.model.element.Modifier;

import static javax.lang.model.element.Modifier.FINAL;
import static javax.lang.model.element.Modifier.PROTECTED;

public final class MarshalSpec {
  private static final TypeName CONTENTVALUES_TYPE =
      ClassName.get("android.content", "ContentValues");
  private static final String CONTENTVALUES_FIELD = "contentValues";
  private static final String CONTENTVALUES_METHOD = "asContentValues";
  private static final String MARSHAL_METHOD_NAME = "marshal";
  private static final String COLUMN_NAME_PARAM = "columnName";
  private static final String CONTENTVALUES_MAP_FIELD = "contentValuesMap";
  private static final ParameterizedTypeName MAP_CLASS =
      ParameterizedTypeName.get(ClassName.get(Map.class), ClassName.get(String.class),
          CONTENTVALUES_TYPE);
  private static final String VALUE = "\"" + SqliteCompiler.KEY_VALUE_VALUE_COLUMN + "\"";

  public static MarshalSpec builder(Table<?> table) {
    return new MarshalSpec(table);
  }

  private final Table<?> table;

  public MarshalSpec(Table<?> table) {
    this.table = table;
  }

  TypeSpec build() {
    TypeSpec.Builder marshal = TypeSpec.classBuilder(table.marshalName())
        .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
        .addTypeVariable(TypeVariableName.get("T", ClassName.get(table.getPackageName(),
            table.interfaceName() + "." + table.marshalName())))
        .addField(table.isKeyValue()
            ? FieldSpec.builder(MAP_CLASS, CONTENTVALUES_MAP_FIELD, FINAL, PROTECTED)
            .initializer("new $T<>()", ClassName.get(LinkedHashMap.class))
            .build()
            : FieldSpec.builder(CONTENTVALUES_TYPE, CONTENTVALUES_FIELD, PROTECTED)
                .initializer("new $T()", CONTENTVALUES_TYPE)
                .build())
        .addMethod(MethodSpec.methodBuilder(CONTENTVALUES_METHOD)
            .addModifiers(Modifier.PUBLIC, FINAL)
            .returns(table.isKeyValue()
                ? ParameterizedTypeName.get(ClassName.get(Collection.class), CONTENTVALUES_TYPE)
                : CONTENTVALUES_TYPE)
            .addStatement(table.isKeyValue() ? "return $L.values()" : "return $L",
                table.isKeyValue()
                    ? CONTENTVALUES_MAP_FIELD
                    : CONTENTVALUES_FIELD)
            .build());

    MethodSpec.Builder constructor = MethodSpec.constructorBuilder()
        .addModifiers(Modifier.PUBLIC);

    for (Column column : table.getColumns()) {
      if (column.isHandledType()) {
        marshal.addMethod(marshalMethod(column));
      } else {
        TypeName columnMarshalType = ClassName.get(table.getPackageName(),
            table.interfaceName() + "." + table.marshalName() + "." + column.marshalName());
        marshal.addType(marshalInterface(column))
            .addField(columnMarshalType, column.marshalField(), Modifier.PRIVATE, FINAL);
        constructor.addParameter(columnMarshalType, column.marshalField())
            .addStatement("this.$L = $L", column.marshalField(), column.marshalField());
        marshal.addMethod(contentValuesMethod(column)
            .addModifiers(Modifier.PUBLIC)
            .returns(TypeVariableName.get("T"))
            .addParameter(column.getJavaType(), column.methodName())
            .addStatement("$L.marshal($L, $L, $L)", column.marshalField(), CONTENTVALUES_FIELD,
                table.isKeyValue() ? VALUE : column.fieldName(), column.methodName())
            .addStatement("return (T) this")
            .build());
      }
    }

    return marshal.addMethod(constructor.build())
        .build();
  }

  private MethodSpec.Builder contentValuesMethod(Column column) {
    if (table.isKeyValue()) {
      MethodSpec.Builder methodBuilder = MethodSpec.methodBuilder(column.methodName());
      if (!column.isNullable() && !column.getJavaType().isPrimitive()) {
        methodBuilder.beginControlFlow("if ($L == null)", column.methodName())
            .addStatement(
                "throw new NullPointerException(\"Cannot insert NULL value for NOT NULL column $L\")",
                column.columnName())
            .endControlFlow();
      }
      return methodBuilder
          .addStatement("$T $L = $L.get($L)", CONTENTVALUES_TYPE, CONTENTVALUES_FIELD,
              CONTENTVALUES_MAP_FIELD, column.fieldName())
          .beginControlFlow("if ($L == null)", CONTENTVALUES_FIELD)
          .addStatement("$L = new $T()", CONTENTVALUES_FIELD, CONTENTVALUES_TYPE)
          .addStatement("$L.put($S, $L)", CONTENTVALUES_FIELD, SqliteCompiler.KEY_VALUE_KEY_COLUMN,
              column.fieldName())
          .addStatement("$L.put($L, $L)", CONTENTVALUES_MAP_FIELD, column.fieldName(),
              CONTENTVALUES_FIELD)
          .endControlFlow();
    }
    return MethodSpec.methodBuilder(column.methodName());
  }

  private TypeSpec marshalInterface(Column column) {
    return TypeSpec.interfaceBuilder(column.marshalName())
        .addModifiers(PROTECTED)
        .addMethod(MethodSpec.methodBuilder(MARSHAL_METHOD_NAME)
            .addParameter(CONTENTVALUES_TYPE, CONTENTVALUES_FIELD)
            .addParameter(ClassName.get(String.class), COLUMN_NAME_PARAM)
            .addParameter(column.getJavaType(), column.methodName())
            .returns(TypeName.VOID)
            .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
            .build())
        .build();
  }

  private MethodSpec marshalMethod(Column column) {
    MethodSpec.Builder method = contentValuesMethod(column)
        .addModifiers(Modifier.PUBLIC)
        .addParameter(column.getJavaType(), column.methodName())
        .returns(TypeVariableName.get("T"));
    if (column.isNullable()) {
      method.beginControlFlow("if ($L == null)", column.methodName())
          .addStatement("$L.putNull($L)", CONTENTVALUES_FIELD,
              table.isKeyValue() ? VALUE : column.fieldName())
          .addStatement("return (T) this")
          .endControlFlow();
    }
    switch (column.type) {
      case INT:
      case LONG:
      case SHORT:
      case DOUBLE:
      case FLOAT:
      case STRING:
      case BLOB:
        method.addStatement("$L.put($L, $L)", CONTENTVALUES_FIELD,
            table.isKeyValue() ? VALUE : column.fieldName(), column.methodName());
        break;
      case BOOLEAN:
        method.addStatement("$L.put($L, $L ? 1 : 0)", CONTENTVALUES_FIELD,
            table.isKeyValue() ? VALUE : column.fieldName(), column.methodName());
        break;
      case ENUM:
        method.addStatement("$L.put($L, $L.name())", CONTENTVALUES_FIELD,
            table.isKeyValue() ? VALUE : column.fieldName(), column.methodName());
        break;
    }
    return method.addStatement("return (T) this")
        .build();
  }
}

File: sqlite-android-compiler/src/main/kotlin/com/alecstrong/sqlite/android/MarshalSpec.kt
package com.alecstrong.sqlite.android

import com.alecstrong.sqlite.android.model.Column
import com.alecstrong.sqlite.android.model.Column.Type
import com.alecstrong.sqlite.android.model.Table
import com.google.common.base.CaseFormat.LOWER_CAMEL
import com.google.common.base.CaseFormat.LOWER_UNDERSCORE
import com.google.common.base.CaseFormat.UPPER_CAMEL
import com.squareup.javapoet.ClassName
import com.squareup.javapoet.FieldSpec
import com.squareup.javapoet.MethodSpec
import com.squareup.javapoet.ParameterizedTypeName
import com.squareup.javapoet.TypeName
import com.squareup.javapoet.TypeSpec
import com.squareup.javapoet.TypeVariableName
import java.util.LinkedHashMap
import javax.lang.model.element.Modifier

import javax.lang.model.element.Modifier.FINAL
import javax.lang.model.element.Modifier.PROTECTED

class MarshalSpec(private val table: Table<*>) {
  internal fun build(): TypeSpec {
    val marshal = TypeSpec.classBuilder(table.marshalName())
        .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
        .addTypeVariable(TypeVariableName.get("T", ClassName.get(table.packageName,
            "${table.interfaceName}.${table.marshalName()}")))

    if (table.isKeyValue) {
      marshal
          .addField(FieldSpec.builder(MAP_CLASS, CONTENTVALUES_MAP_FIELD, FINAL, PROTECTED)
              .initializer("new ${'$'}T<>()", ClassName.get(LinkedHashMap::class.java))
              .build())
          .addMethod(MethodSpec.methodBuilder(CONTENTVALUES_METHOD)
              .addModifiers(Modifier.PUBLIC, FINAL)
              .returns(ParameterizedTypeName.get(ClassName.get(Collection::class.java),
                  CONTENTVALUES_TYPE))
              .addStatement("return $CONTENTVALUES_MAP_FIELD.values()")
              .build())
    } else {
      marshal
          .addField(FieldSpec.builder(CONTENTVALUES_TYPE, CONTENTVALUES_FIELD, PROTECTED)
              .initializer("new ${'$'}T()", CONTENTVALUES_TYPE)
              .build())
          .addMethod(MethodSpec.methodBuilder(CONTENTVALUES_METHOD)
              .addModifiers(Modifier.PUBLIC, FINAL)
              .returns(CONTENTVALUES_TYPE)
              .addStatement("return $CONTENTVALUES_FIELD")
              .build())
    }

    val constructor = MethodSpec.constructorBuilder().addModifiers(Modifier.PUBLIC)

    for (column in table.columns) {
      if (column.isHandledType) {
        marshal.addMethod(marshalMethod(column))
      } else {
        val columnMarshalType = ClassName.get(table.packageName,
            "${table.interfaceName}.${table.marshalName()}.${column.marshalName()}")
        marshal.addType(marshalInterface(column))
            .addField(columnMarshalType, column.marshalField(), Modifier.PRIVATE, FINAL)
        constructor.addParameter(columnMarshalType, column.marshalField())
            .addStatement("this.${column.marshalField()} = ${column.marshalField()}")
        marshal.addMethod(contentValuesMethod(column)
            .addModifiers(Modifier.PUBLIC)
            .returns(TypeVariableName.get("T"))
            .addParameter(column.javaType, column.methodName)
            .addStatement("${column.marshalField()}.marshal($CONTENTVALUES_FIELD, " +
                "${column.name()}, ${column.methodName})")
            .addStatement("return (T) this")
            .build())
      }
    }

    return marshal.addMethod(constructor.build()).build()
  }

  private fun contentValuesMethod(column: Column<*>) =
      if (table.isKeyValue) {
        val methodBuilder = MethodSpec.methodBuilder(column.methodName)
        if (!column.isNullable && !column.javaType.isPrimitive) {
          methodBuilder.beginControlFlow("if (${column.methodName} == null)")
              .addStatement("throw new NullPointerException(" +
                  "\"Cannot insert NULL value for NOT NULL column ${column.name}\")")
              .endControlFlow()
        }
        methodBuilder
            .addStatement("${'$'}T $CONTENTVALUES_FIELD = " +
                "$CONTENTVALUES_MAP_FIELD.get(${column.fieldName})", CONTENTVALUES_TYPE)
            .beginControlFlow("if ($CONTENTVALUES_FIELD == null)")
            .addStatement("$CONTENTVALUES_FIELD = new ${'$'}T()", CONTENTVALUES_TYPE)
            .addStatement("$CONTENTVALUES_FIELD.put(${'$'}S, ${column.fieldName})",
                SqliteCompiler.KEY_VALUE_KEY_COLUMN)
            .addStatement("$CONTENTVALUES_MAP_FIELD.put(${column.fieldName}, $CONTENTVALUES_FIELD)")
            .endControlFlow()
      } else MethodSpec.methodBuilder(column.methodName)

  private fun marshalInterface(column: Column<*>) =
      TypeSpec.interfaceBuilder(column.marshalName())
          .addModifiers(PROTECTED)
          .addMethod(MethodSpec.methodBuilder(MARSHAL_METHOD_NAME)
              .addParameter(CONTENTVALUES_TYPE, CONTENTVALUES_FIELD)
              .addParameter(ClassName.get(String::class.java), COLUMN_NAME_PARAM)
              .addParameter(column.javaType, column.methodName)
              .returns(TypeName.VOID)
              .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
              .build())
          .build()

  private fun marshalMethod(column: Column<*>) =
      if (column.isNullable && (column.type == Type.ENUM || column.type == Type.BOOLEAN)) {
        contentValuesMethod(column)
            .beginControlFlow("if (${column.methodName} == null)")
            .addStatement("$CONTENTVALUES_FIELD.putNull(${column.fieldName})")
            .addStatement("return (T) this")
            .endControlFlow()
      } else {
        contentValuesMethod(column)
      }
          .addModifiers(Modifier.PUBLIC)
          .addParameter(column.javaType, column.methodName)
          .returns(TypeVariableName.get("T"))
          .addStatement("$CONTENTVALUES_FIELD.put(${column.name()}, ${column.marshaledValue()})")
          .addStatement("return (T) this")
          .build()

  private fun Table<*>.marshalName() = name + "Marshal"
  private fun Column<*>.name() = if (table.isKeyValue) VALUE else fieldName
  private fun Column<*>.marshalName() = LOWER_UNDERSCORE.to(UPPER_CAMEL, name) + "Marshal";
  private fun Column<*>.marshalField() = LOWER_UNDERSCORE.to(LOWER_CAMEL, name) + "Marshal";
  private fun Column<*>.marshaledValue() =
      when (type) {
        Column.Type.INT, Column.Type.LONG, Column.Type.SHORT, Column.Type.DOUBLE, Column.Type.FLOAT,
        Column.Type.STRING, Column.Type.BLOB -> methodName
        Column.Type.BOOLEAN -> "$methodName ? 1 : 0"
        Column.Type.ENUM -> "$methodName.name()"
        else -> throw IllegalStateException("Unexpected type")
      }

  companion object {
    private val CONTENTVALUES_TYPE = ClassName.get("android.content", "ContentValues")
    private val CONTENTVALUES_FIELD = "contentValues"
    private val CONTENTVALUES_METHOD = "asContentValues"
    private val MARSHAL_METHOD_NAME = "marshal"
    private val COLUMN_NAME_PARAM = "columnName"
    private val CONTENTVALUES_MAP_FIELD = "contentValuesMap"
    private val MAP_CLASS = ParameterizedTypeName.get(ClassName.get(Map::class.java),
        ClassName.get(String::class.java),
        CONTENTVALUES_TYPE)
    private val VALUE = "\"" + SqliteCompiler.KEY_VALUE_VALUE_COLUMN + "\""

    internal fun builder(table: Table<*>) = MarshalSpec(table)
  }
}
